kvserver: only pause followers when holding active lease

tbg opened the pull request:
If the raft leader is not the leaseholder (which includes the case in
which we just transferred the lease away), leave all followers unpaused.
Otherwise, the leaseholder won't learn that the entries it submitted
were committed which effectively causes range unavailability.

Fixes #84884.

Release note: None

tbg committed:
kvserver: extract method updatePausedFollowersLocked

Intentionally leaving this unsimplified as a purely mechanical commit.

Release note: None

tbg committed:
kvserver: simplify updatePausedFollowersLocked

Release note: None

tbg committed:
kvserver: only pause followers when holding active lease

If the raft leader is not the leaseholder (which includes the case in
which we just transferred the lease away), leave all followers unpaused.
Otherwise, the leaseholder won't learn that the entries it submitted
were committed which effectively causes range unavailability.

Fixes #84884.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/85732)
<!-- Reviewable:end -->

tbg commented on pkg/kv/kvserver/replica_raft_overload.go:
```diff
@@ -206,3 +206,73 @@ func (osm *overloadedStoresMap) Swap(
+	}
+
+	status := r.leaseStatusAtRLocked(ctx, r.Clock().NowAsClockTimestamp())
+	if !status.IsValid() || status.OwnedBy(r.StoreID()) {
```
missing a `!` here, caught by the test

tbg commented:
Note: also still need to make sure a leader never considers itself paused.

tbg committed:
kvserver: never pause local replica

As observed in #84884, an overloaded store that held leases could end up
"pausing" replication traffic to itself. This (likely) had no practical
effect since the leader never sends messages to itself, but it meant
reporting bogus counts of paused replicas.

This commit ensures that a raft leader will never pause itself.

Release note: None

erikgrinaker commented on pkg/kv/kvserver/replica_raft_overload.go:
```diff
@@ -206,3 +206,74 @@ func (osm *overloadedStoresMap) Swap(
+		return
+	}
+
+	status := r.leaseStatusAtRLocked(ctx, r.Clock().NowAsClockTimestamp())
+	if !status.IsValid() || !status.OwnedBy(r.StoreID()) {
+		// If we're not the leaseholder (which includes the case in which we just
+		// transferred the lease away), leave all followers unpaused. Otherwise, the
+		// leaseholder won't learn that the entries it submitted were committed
+		// which effectively causes range unavailability.
+		return
+	}
```
Unpausing if there's no valid lease seems like a prudent failsafe.

However, it seems a bit lenient to unpause all followers when we don't have the lease. There might be cases where this situation persists, or where it causes intermittent unpausing. Should we only exclude the leaseholder instead, or are we worried that might open for other pathologies?

tbg replied:
Really hard to say without the benefit of hindsight. But I'd much rather approach pausing from below (adding complexity only when proven relevant) vs trying to anticipate problems and then (perhaps painfully) finding out that there was something we missed.
I expect it to be easier to see that a node is still sending to an overloaded follower (esp. in light of the new admission control logs that are landing ~right about now) vs figuring out that some loss of quorum or other degradation is caused by pausing.

tbg commented:
bors r=erikgrinaker
TFTR!
